### ISSUE TITLE
Proposal: Deprecation of cookie splitting

### ISSUE THREAD

**Issue by JoelSpeed** [oauth2-proxy/oauth2-proxy#482]
Currently (v5.1.0), when a session is encrypted and the size of the session exceeds 4kb, there is logic within the proxy to [split the cookie](https://github.com/oauth2-proxy/oauth2-proxy/blob/7efc162aaa1a8cbd540fe1d2dab2106ddd8278d4/pkg/sessions/cookie/session_store.go#L128-L153). This is required so that we do not exceed the maximum size of a cookie. As far as I'm aware this is guaranteed to happen when using Azure because of the size of the ID Tokens they generate, and it has caused a number of issues in the past.

There is an alternate solution however. By using redis as a session store (potentially we need to implement other alternate session stores? sqlite could be a good option for small deployments, dynamo or other cloud dbs for scaled deployments? Kubernetes CRDs for those running on Kubernetes), the larger sessions can be stored server side and the client need only keep track of a session ticket encrypted within the cookie, which is typically much smaller.

There are other additional benefits of server side session storage as well, eg you are less likely to have authentication fail during refreshes as the session can not go stale in a cookie.

My proposal would be that we remove this behaviour from the proxy in a couple of steps:
- First, introduce a warning that the cookie is being split
- Secondly either remove the splitting altogether or feature gate it behind a flag
    - If feature gated, make a note that this is unsupported behaviour and that support will not be provided when using the feature

I would like to gather feedback from the community on this idea before we make a breaking change, so please post comments below

Related issues:
- #458 
- #350
- #138 
- #69 
- #29 
- #388 
- #266

**Comment by steakunderscore** [comment]
I'm all for this change. I agree this is something that should for a few months/releases be a deprecation warning.

**Comment by johejo** [comment]
Agree.
Large cookies are not a good choice as they may be limited by the L7 load balancer (PaaS).
Recently I faced the HTTP/2 enabled Azure Application Gateway blocks large header (it means large cookie).
https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#application-gateway-limits

It may be possible to support full local boot modes other than Redis (eg in-memory session storage).
They don't share storage when scaled horizontally and should be warned because they can't be used in production environments, but I think it's a good option for standalone use cases.

Also, sqlite3 driver for Go requires CGO, so some people might not like it.
https://github.com/mattn/go-sqlite3

**Comment by JoelSpeed** [comment]
Thanks for your thoughts @johejo :)

I think we should come up with a deprecation plan for this, we currently have a warning in the proxy from the next release that warns that the cookies are being split, should we mark this as deprecated in the same release or wait for a patch release to do that?

 >Also, sqlite3 driver for Go requires CGO, so some people might not like it.
https://github.com/mattn/go-sqlite3

I wonder if there are other ways to achieve this that wouldn't need CGO.

Alternative suggestions for potential storage implementations for sessions:
- SQL Driver for MySQL/Postgres (potentially overkill?)
- Kubernetes CRD storage (like Dex does)
- Memory (for testing purposes?)
- Etcd (or other popular Key Value stores)
- Cloud Provider Key/Value stores (eg AWS DynamoDB)

**Comment by NickMeves** [comment]
Something on my mind after I looked at tuning up the redis session store that I might take a stab at reorganizing:

We could potentially split out all common logic dealing with encrypting the data, generated a session ticket for the cookie, and the like -- and make that common for all non-cookie session stores.

Then any new session store interfaces just need to know how to initialize and implement the Save/Load/Clear functionalities with a generic key/value (with the common logic dealing with making a key, encrypting, reading the cookie, etc).

Should make it easier to add in more session store types easily.

**Comment by JoelSpeed** [comment]
@NickMeves I had similar thoughts before, and agree, would definitely make implementing different session stores easier. Seems like a reasonable refactoring project to me

**Comment by JoelSpeed** [comment]
I think we should add a deprecation warning for the next minor release. This has unfortunately not seen much traction from the community though, I'd like to provide server side implementations for the session storage but not sure which would actually be valuable.

**Comment by NickMeves** [comment]
So I'm actually a big fan of the cookie session store (and a little bit the auto-splitting as well). For these reasons:
- Ease of setup / Barrier to entry is super low.  No extra architectural requirements. Especially for deployments where oauth2-proxies are deployed everywhere as sidecar containers to web app pods, the standalone session management is ideal.
- Gives the option for better performance/less architectural points of failure. Removes a network hop from from proxy -> session store that would be added for every request & uptime requirement of the session store.

I remember when doing a POC of potential access proxies at the beginning of this year that this ease of getting a session managed in a cookie made oauth2-proxy stand out from ORY Oathkeeper.

Cookie splitting is potentially a nice feature when used appropriately to preserve rich/verbose OIDC tokens (when used appropriately 😉  ) -- IE splitting and only being at 5-6KB cookie size is perfectly fine.  This seems to rear its ugly head when users bump into 8KB header limit in some environments (nginx ingress in some node types & azure seem to be the big offenders).

Do we potentially need to be more diligent about warning about approaching 8KB headers to give users the heads up the true cause of the issue? (cookie splitting obviously being a big contributor to that when cookie size approaches a large value).

So maybe making the warning in the logs less about cookie splitting and more about session size being dangerously large for cookie session store?

**Comment by NickMeves** [comment]
> I think we should add a deprecation warning for the next minor release. This has unfortunately not seen much traction from the community though, I'd like to provide server side implementations for the session storage but not sure which would actually be valuable.

I would think K8s based CRD session store might be the next potential best addition?

Pros:
- Easy setup.  Transparent architecturally to end users in K8s ecosystem without additional deployment/statefulsets as session store dependencies. Great easy alternative when cookie session store breaks down.

Cons:
- Might be slow. Relies on etcd for storage under the hood -- so disk based and raft consensus algorithm considerations. Vs Redis where updates live in memory (and syncing to disk appendonly.aof and replica node are trailing and non-blocking from operations on the primary node).
- Potential to roll over an etcd cluster and take out a K8s cluster if it gets too big (too many users/sessions). I haven't dealt with this myself but I've heard of K8s admins rolling out 2 etcd (one for kube-system one for elsewhere) to prevent this scenario.

**Comment by kfox1111** [comment]
Deploying/managing Redis is significantly more work then just cookie splitting.

Cookie splitting is also great insurance for when your usually under the limit but every once in a while it may peak over the limit. This prevents random seeming breakages for users. Having support for splitting + issuing prometheus metrics for when it happens would allow admins to deploy without the need of redis, and once usage starts to get too big, then there is plenty of time to learn how to deploy redis,

Please do keep splitting. Its a great feature I think.

**Comment by kfox1111** [comment]
> Cookie splitting is potentially a nice feature when used appropriately to preserve rich/verbose OIDC tokens (when used appropriately  ) -- IE splitting and only being at 5-6KB cookie size is perfectly fine. This seems to rear its ugly head when users bump into 8KB header limit in some environments (nginx ingress in some node types & azure seem to be the big offenders).
> 
> Do we potentially need to be more diligent about warning about approaching 8KB headers to give users the heads up the true cause of the issue? (cookie splitting obviously being a big contributor to that when cookie size approaches a large value).

Making this a prometheus metric might help too. Maximum header size returned in requests.

**Comment by JoelSpeed** [comment]
> I would think K8s based CRD session store might be the next potential best addition? 

I would certainly be in favour of this, it feels like a natural fit and it would be relatively easy to maintain personally since I write K8s code for my day job 😅 I think we have two options if we go down this route, we either store the sessions in secrets within Kubernetes, which would be good since they're built in, but have the limitation of people maybe having too wide RBAC permissions so access to them may not be that secure, or we create our own CRDs, this would be more work, but would mean that we are less likely for people to have overly scoped permissions allowing them to potentially modify the stored session data.

The one thing we would have to be careful about here is garbage collection. Somehow we will need to ensure that expired sessions are cleaned up.

> Deploying/managing Redis is significantly more work then just cookie splitting.

Do you think alternative stores would be helpful in this scenario, eg if you ran on AWS, would having a Dynamo store option make that an easy was for you to have server side storage? Or would you still lean towards a cookie based storage?

**Comment by JoelSpeed** [comment]
> Cookie splitting is potentially a nice feature when used appropriately to preserve rich/verbose OIDC tokens (when used appropriately 😉 ) -- IE splitting and only being at 5-6KB cookie size is perfectly fine. This seems to rear its ugly head when users bump into 8KB header limit in some environments (nginx ingress in some node types & azure seem to be the big offenders).

There is also another issue with the cookie splitting, in that NGINX by default will only copy the first `Set-Cookie` header from a subrequest, so to be able to use nginx auth request with the split cookies, you have to install a lua module and use a lua script one of our contributors came up with, easy for those running nginx as an ingress controller, not so easy for those who don't run K8s

**Comment by NickMeves** [comment]
> There is also another issue with the cookie splitting, in that NGINX by default will only copy the first `Set-Cookie` header from a subrequest, so to be able to use nginx auth request with the split cookies, you have to install a lua module and use a lua script one of our contributors came up with, easy for those running nginx as an ingress controller, not so easy for those who don't run K8s

Is this still the case? I just started experimenting with nginx-ingress annotations yesterday on a dummy httpbin deployment's ingress.

I just had to set the `nginx.ingress.kubernetes.io/auth-signin` and `nginx.ingress.kubernetes.io/auth-url` annotations. I saw my cookie session (and split to boot) in the httpbin response.

I haven't dived into the nginx ingress implementation of oauth auth -- but I wonder if they handle the cookies from the subrequest for us now?

**Comment by NickMeves** [comment]
Yup, just tested again to confirm. httpbin backend got my giant 6KB split cookie without any added Lua 👍

**Comment by JoelSpeed** [comment]
Interesting, it certainly never used to as we had a number of people come to us with this problem, but if it's fixed now, that's great news!

**Comment by NickMeves** [comment]
> I would certainly be in favour of this, it feels like a natural fit and it would be relatively easy to maintain personally since I write K8s code for my day job 😅 I think we have two options if we go down this route, we either store the sessions in secrets within Kubernetes, which would be good since they're built in, but have the limitation of people maybe having too wide RBAC permissions so access to them may not be that secure, or we create our own CRDs, this would be more work, but would mean that we are less likely for people to have overly scoped permissions allowing them to potentially modify the stored session data.
> 
> The one thing we would have to be careful about here is garbage collection. Somehow we will need to ensure that expired sessions are cleaned up.

I actually don't think we need to be worried about the data being in a secret vs a CRD for instance. If the session Ticket logic I'm trying to refactor is used for every store https://github.com/oauth2-proxy/oauth2-proxy/pull/682 -- then the Ticket.Secret provides us a unique per session AES-GCM secret that is distinct for each session.

All secret would get us then is double AES encryption on that data I think? (assuming the K8s cluster implemented encryption at rest in etcd for secrets).

> 
> > Deploying/managing Redis is significantly more work then just cookie splitting.
> 
> Do you think alternative stores would be helpful in this scenario, eg if you ran on AWS, would having a Dynamo store option make that an easy was for you to have server side storage? Or would you still lean towards a cookie based storage?

Interesting, the cloud vendors generic key/value stores would be an interesting proposition that might ease the barrier of entry. Good idea!

I don't know how founded my worry is, but for all these the session retrieval needing an extra network hop to the session store had me worried about high performance use cases. I haven't measured how much the penalty is though. But for those scenarios I imagine cookie sessions is the most desireable.

**Comment by kfox1111** [comment]
> I would certainly be in favour of this, it feels like a natural fit and it would be relatively easy to maintain personally since I write K8s code for my day job  I think we have two options if we go down this route, we either store the sessions in secrets within Kubernetes, which would be good since they're built in, but have the limitation of people maybe having too wide RBAC permissions so access to them may not be that secure, or we create our own CRDs, this would be more work, but would mean that we are less likely for people to have overly scoped permissions allowing them to potentially modify the stored session data.

Secrets would be a next best option I think, yeah. They do support disk encryption at rest in the apiserver.

> The one thing we would have to be careful about here is garbage collection. Somehow we will need to ensure that expired sessions are cleaned up.

+1

> Do you think alternative stores would be helpful in this scenario, eg if you ran on AWS, would having a Dynamo store option make that an easy was for you to have server side storage? Or would you still lean towards a cookie based storage?

No. My users and I usually use oauth2-proxy on bare metal private clusters. So the bar goes from nothing to maintaining a db yourself currently all in one go.

**Comment by JoelSpeed** [comment]
> I don't know how founded my worry is, but for all these the session retrieval needing an extra network hop to the session store had me worried about high performance use cases. I haven't measured how much the penalty is though. But for those scenarios I imagine cookie sessions is the most desirable.

This is where we probably want to talk about caching the responses from these stores, we know when the session is due to be expired or refreshed, so we could cache it for a short period so long as we don't overstay that period, though, that's a separate discussion.

> I actually don't think we need to be worried about the data being in a secret vs a CRD for instance. 

I'm more concerned about the overhead of managing our own CRD vs using a built-in. Whether it's sercet or configmap it doesn't particularly matter as you say, though I'd lean towards secret

> No. My users and I usually use oauth2-proxy on bare metal private clusters. So the bar goes from nothing to maintaining a db yourself currently all in one go.

With a little more background, your motivation here makes more sense 😅 Ok, that's interesting to know then. Do you run baremetal K8s in these environments?

One other question on my mind, is whether an etcd store would be good, since people with K8s are running one already, is it hard to run a second or to share the etcd? It would save some of the overhead of going via the K8s API server

**Comment by kfox1111** [comment]
> This is where we probably want to talk about caching the responses from these stores, we know when the session is due to be expired or refreshed, so we could cache it for a short period so long as we don't overstay that period, though, that's a separate discussion.

Yeah, you could watch the kubernetes objects with a k8s watch and cache them.

> With a little more background, your motivation here makes more sense  Ok, that's interesting to know then. Do you run baremetal K8s in these environments?

Yeah, mostly bare metal k8s. We could explore an operator or something for maintaining the state in a clustered database. But really, the more things to manage, the more things can go wrong, and or need to educate folks on how to maintain. (even if an operator is mostly doing it). Its nice to keep things simple until more is really required. But have an option to smoothly transition when required rather then have it break and have to rush.

> One other question on my mind, is whether an etcd store would be good, since people with K8s are running one already, is it hard to run a second or to share the etcd? It would save some of the overhead of going via the K8s API server

I heard the etcd operator is deprecated. None of my users have raw access to the etcd maintaining k8s. That would be a big no no. So they can't run their own oauth2-proxy pointed at an etcd easily.

K8s storage would be much easier on users I think then asking for an external db such as redis, mongodb, percona, etcd, etc. When running in k8s, it really should be a clustered database to fit its upgrade model. They're complicated so ideally shouldn't be needed until you reach a certain scale of usage. Having a single user with a few dozen groups shouldn't be enough I think to justify that kind of requirement? Just thinking out loud...

**Comment by NickMeves** [comment]
So as a data point for a CRD instead of a traditional resource like a Secret or ConfigMap:

Recently we had some dev test system leaking stale `secret` resources (to the tune of 5000+ over time in one K8s cluster). This only became painfully apparently when we migrated to Helm3 -- since it manages it revisions in Secrets now.

Performing a `helm list` on the namespace with the secret bonanza was like a 2+ minute operation.

So with something like sessions where we are potentially storing 100s-1000s depending on the user population -- this might make K8s admins live's a nightmare when they want to manage normal versions of the resource we pick to store sessions.

Hopefully `helm list`'s apparently O(N) behavior was an issue with their implementation and getting K8s resources is O(logN) or O(1)... otherwise any K8s native store might be a no-go.

**Comment by NickMeves** [comment]
Another crazy idea I had for a session store last night:

Shared (needs a cluster bus) in-memory session cache using something like [groupcache](https://github.com/golang/groupcache) [https://github.com/golang/groupcache] (likely [mailgun's fork](https://github.com/mailgun/groupcache) that supports key removal & expiration [https://github.com/mailgun/groupcache])

With maybe [memberlist](https://github.com/hashicorp/memberlist) to setup a cluster bus over `/oauth2/cluster` for instance.

Has some complex details though, like startup syncing with existing caches on up nodes (otherwise K8s cluster node rotations would slowly wipe out an existing distributed cache)

Outside of this idea -- a cluster bus might be a solution to the refresh singleflight discussion here: https://github.com/oauth2-proxy/oauth2-proxy/issues/667

**Comment by JoelSpeed** [comment]
> Yeah, you could watch the kubernetes objects with a k8s watch and cache them.

Definitely, we can use an informer for the K8s implementation, but it would be good to also have some caching for the redis sessions as well maybe?

>Recently we had some dev test system leaking stale secret resources (to the tune of 5000+ over time in one K8s cluster). This only became painfully apparently when we migrated to Helm3 -- since it manages it revisions in Secrets now.

I think if we went with `sercets`, which would be my preference, then we would recommend that a separate namespace is used just for the OAuth2 Proxy or its secrets at the very least. We should have 1 session per user at any given time, we will have to implement a garbage collector to make sure stale sessions are removed. This will need thorough testing. Since we only want to have one field, it seems like a lot of work to have a CRD, unless we didn't use the same ticket style as with redis and didn't encrypt and compress, in which case we would want to use a secret for encryption at rest anyway 🤷 

> Shared (needs a cluster bus) in-memory session cache using something like groupcache [https://github.com/golang/groupcache] (likely mailgun's fork that supports key removal & expiration [https://github.com/mailgun/groupcache])

This is an interesting idea, though I worry that might expand the scope of the project quite drastically 🤔 Also, would you lose all sessions if you restarted all proxies at once?

**Comment by kfox1111** [comment]
separate namespace for secrets would be nice as an option, but some users don't have access to multiple namespaces, just one.

**Comment by JoelSpeed** [comment]
Can you give an example of where that might be the case? It would be my assumption that if a user has access just to a single namespace, that they would not be able to install a CRD since that is a cluster scoped resource?

**Comment by kfox1111** [comment]
@JoelSpeed that is correct. CRD installation is not issued to users with a rolebinding of 'admin'. Only cluster admins can do that.

additionally, if you install the operator-lifecycle-manager(OLM) on your cluster, the OLM is allowed to install CRD's on behalf of users that subscribe to an Operator (request it be installed) if the CRD is not already installed. So that can be an alternate mechanism to get CRD's installed on the cluster.

But in general, CRD's can't be installed on a cluster without a cluster-admin's approval or delegation.

**Comment by NickMeves** [comment]
> This is an interesting idea, though I worry that might expand the scope of the project quite drastically 🤔 Also, would you lose all sessions if you restarted all proxies at once?

Oh yeah - definitely 😄  -- If something like this were ever a session store someone picked, they'd need to be really comfortable with their PodDisruptionBudget setup. Which is why resyncing on startup before `ping` returning healthy would be key otherwise a rolling node restart of a cluster would likely wipe alot of sessions even with a PDB.

I had the crazy idea the other night -- then decided I'm likely going with Redis myself for a new rollout I'm doing where I can't have the big cookies. We can let the idea live on here in case it inspires someone else to try implementing it.

**Comment by NickMeves** [comment]
> I think if we went with `sercets`, which would be my preference, then we would recommend that a separate namespace is used just for the OAuth2 Proxy or its secrets at the very least. We should have 1 session per user at any given time, we will have to implement a garbage collector to make sure stale sessions are removed. This will need thorough testing. Since we only want to have one field, it seems like a lot of work to have a CRD, unless we didn't use the same ticket style as with redis and didn't encrypt and compress, in which case we would want to use a secret for encryption at rest anyway 🤷

I imagine using `oauth2-proxy` as a sidecar container in app pods is a popular deployment strategy (it gives the most security benefits since the app container can be locked to listen on localhost only preventing auth header injection attacks by circumventing the proxy).

Would the separate `secret` namespace work with that deployment design? Since `oauth2-proxy` itself is then firmly in the app namespace?

**Comment by JoelSpeed** [comment]
I don't see why it couldn't work across namespaces. You can grant permissions for a service account to have access to a different namespace, so there's nothing preventing you from doing that, all we need from oauth proxy perspective is the target namespace, and the credentials

**Comment by kfox1111** [comment]
Remember there is a difference between users of a cluster and administrators of a cluster. You only can setup cross namespace permissions if you have permissions to 2 namespaces. Some cluster-admins will not give out multiple namespaces to users. Its not a technical restriction, but a restriction none the less. Another example, some users may get charged by the cluster-admin per namespace. So there are multiple reasons for supporting same namespace even though it may not be ideal. Supporting separate namespaces optionally though is good.

**Comment by kfox1111** [comment]
@NickMeves While we are talking about all the cool things you can do with Kubernetes, (sidecars)
A mutating webhook would be cool too. You could do oauth2-proxy injection as a sidecar automatically based on a pod annotation similar to how istio does.

**Comment by NickMeves** [comment]
Is this still an issue now that our cookies are LZ4 compressed by default now? I think even our former biggest offenders (Azure) are under the 4K limit now in normal circumstances.

**Comment by JoelSpeed** [comment]
I'm happy to close this for now, we have seen a reduction in the number of complaints and confusions over the cookie split feature

**Comment by andoks** [comment]
@JoelSpeed: could this issue benefit from being re-opened? When investigating the best way of handling too large cookies, I have gotten the impression that the [cookie-splitting method has a tendency of resulting in undesired behavior](https://github.com/oauth2-proxy/oauth2-proxy/issues/1940#issuecomment-1364008273), and if I could, I would want to avoid to have to spin up Redis. 

Or perhaps a new Proposal for an alternative on-disk cookie-store (e.g by using Sqlite as previously suggested - if a pure Go implementation is preferred, maybe [modernc.org/sqlite](https://pkg.go.dev/modernc.org/sqlite) could be an alternative, or something like [bbolt](https://github.com/etcd-io/bbolt) perhaps) for single-instance deployments?

Sorry for necroing this old issue if not relevant.

Related links:
- https://github.com/oauth2-proxy/oauth2-proxy/issues/1940#issuecomment-1364008273
- https://github.com/oauth2-proxy/oauth2-proxy/issues/644
- https://github.com/oauth2-proxy/oauth2-proxy/pull/632
- https://github.com/oauth2-proxy/oauth2-proxy/issues/676#issuecomment-657129910
- [github search for "WARNING: Multiple cookies are required...](https://github.com/search?q=repo%3Aoauth2-proxy%2Foauth2-proxy+WARNING%3A+Multiple+cookies+are+required+for+this+session+as+it+exceeds+the+4kb+cookie+limit&type=issues)

**Comment by JoelSpeed** [comment]
> K8s and PostgreSQL would be among my favorite.

PRs welcome if anyone has the time

**Comment by dastrobu** [comment]
I was looking into an implementation of a PostgreSQL session store.

There are some challenges, though:

1. PostgreSQL does not have an expiration time-based retention for rows. That would require a workaround, for example, cleaning up expired sessions from a scheduled job using the pg_cron extension.  
2. PostgreSQL does not have a lock comparable to the Redis lock.

So I ended up solving the cookie splitting problem in a different way (in a fork of this project).

As the ID token consumes quite some space in the session cookie, I added another flag: `session-cookie-no-id-token`, which allows us to skip storing the ID token in the session, as only the access token and refresh tokens are required upstream.

This was inspired by the already available `session-cookie-minimal` flag, which skips storing all tokens in the session. This is too extreme, as we need the access tokens and refresh tokens to work.

The optimization of not storing the ID token was sufficient to prevent the cookie splitting from happening.

Here is the relevant code:

```go
func (s *SessionStore) cookieForSession(ss *sessions.SessionState) ([]byte, error) {
	if s.Minimal && (ss.AccessToken != "" || ss.IDToken != "" || ss.RefreshToken != "") {
		minimal := *ss
		minimal.AccessToken = ""
		minimal.IDToken = ""
		minimal.RefreshToken = ""

		return minimal.EncodeSessionState(s.CookieCipher, true)
	}

	if s.NoIDToken && (ss.IDToken != "") {
		withoutIDToken := *ss
		withoutIDToken.IDToken = ""
		return withoutIDToken.EncodeSessionState(s.CookieCipher, true)
	}

	return ss.EncodeSessionState(s.CookieCipher, true)
}
```

I can open a proper PR including tests and docs if that is a desired feature on this upstream project.

**Comment by gempir** [comment]
In a big mature deployment I would always use something like Redis or a SQL DB.

But oauth2-proxy is used for hobby projects, or just side projects that make the deployment of that so much more complicated.

I really like the suggestions of an in-memory store. Go has great tools to build such a thing relatively easily and simple.

It could facilitate removing the cookie option, the only painpoint you have is when you have a HA deployment on multiple servers or you restart the process frequently.

Which IMO are uncommon in those smaller hobby scenarios.
